EpochIOGroup

dianarg opened the pull request:
- Fixes #1125

dianarg committed:
Add EpochIOGroup

Change-Id: I21c8a96a19a545f0ae136d5fb5afad177630e6d6
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add static methods for event type string conversion in ApplicationSampler

Change-Id: Id1e938937fe8154cbffd1af2ce9f18279cbb552d
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add cpu-process mapping to ApplicationSampler

Change-Id: I5ca5fe0163a025ae38ebdc677d11f4ab9c71ebe9
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add integration tests for EpochIOGroup based on record stream

Change-Id: I91061c97982bdb36d86ebe0405bfd43ffd6336b7
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Fix ApplicationSampler filter construction

Change-Id: Ie65b1728b348e1269c3506c5c29c3d7df147d650
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Handle NAN for aggregated signals

Change-Id: Id1aa65468040d40af279148635d66842c8ca9dca
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add EpochIOGroup to PlatformIO

Change-Id: I10af48fa066ebce0148b4d692c3dd634148d0378
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo commented on src/ApplicationSampler.cpp:
```diff
@@ -267,6 +301,11 @@ namespace geopm
         return {};
     }
 
+    std::vector<int> ApplicationSamplerImp::per_cpu_process_id(void) const
```
per_cpu_process()

cmcantalupo commented on src/Agg.cpp:
```diff
@@ -122,45 +136,49 @@ namespace geopm
             sum_squared *= sum_squared;
-            std::vector<double> operand_squared(operand);
-            for (auto &it : operand_squared) {
+            std::vector<double> filtered_squared(filtered);
```
filtered_squared -> operand_squared

cmcantalupo commented on src/ApplicationSampler.hpp:
```diff
@@ -145,6 +145,9 @@ namespace geopm
             virtual std::vector<double> per_cpu_progress(void) const = 0;
+            /// @brief Return a per-cpu vector of the process mapped
+            ///        to each cpu.
+            virtual std::vector<int> per_cpu_process_id(void) const = 0;
```
same, remove "_id"

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+        if (result == -1) {
+            result = m_active_signal.size();
+            m_active_signal.push_back({signal_type, domain_type, domain_idx});
+            // m_is_signal_pushed = true;
```
remove dead code

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+        auto records = m_app.get_records();
+        for (const auto &record : records) {
+            if (m_epoch_map.find(record.process) == m_epoch_map.end()) {
+                /// @todo: this should not happen; for now skip record
```
add this check as a debug assert.

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+                /// @todo: this should not happen; for now skip record
+                continue;
+            }
+            m_epoch_map.at(record.process)->update(record);
```
keep a local copy of the m_epoch_map iterator and check if process hasn't changed before doing map look up. Of questionable value.  Use godbolt and don't be a smart ass.

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+
+    std::unique_ptr<IOGroup> EpochIOGroup::make_plugin(void)
+    {
+        return std::unique_ptr<IOGroup>(new EpochIOGroup);
```
return geopm::make_unique<EpochIOGroup>();

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+            case M_SIGNAL_EPOCH_RUNTIME:
+            case M_SIGNAL_EPOCH_RUNTIME_NETWORK:
+            case M_SIGNAL_EPOCH_RUNTIME_IGNORE:
+                break;
```
Add comment that default Agg::max is set above

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+            case M_SIGNAL_EPOCH_RUNTIME:
+            case M_SIGNAL_EPOCH_RUNTIME_NETWORK:
+            case M_SIGNAL_EPOCH_RUNTIME_IGNORE:
+                break;
```
same as above

cmcantalupo commented on src/EpochIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+                result = "Last runtime between epoch events sampled from the process on the given CPU";
+                break;
+            case M_SIGNAL_EPOCH_RUNTIME_NETWORK:
+                result = "Portion of the last epoch with the network hint sampled from the process on the given CPU";
```
Portion of last runtime between epoch event where the application provided the network hint.

cmcantalupo commented on src/IOGroup.cpp:
```diff
@@ -63,6 +64,8 @@ namespace geopm
                         TimeIOGroup::make_plugin);
         register_plugin(CpuinfoIOGroup::plugin_name(),
                         CpuinfoIOGroup::make_plugin);
+        register_plugin(EpochIOGroup::plugin_name(),
```
Comment about the temporary override imposed by the ordering here

cmcantalupo reviewed (commented):
Just a few little things

dianarg committed:
Address review comments

- TODO: investigate performance optimization for map lookup with
  processing records in read_batch()

Change-Id: I8572777ae38042218b8d813f3a30752ff4672b99
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
